ci/date-check/src/main.rs

```diff
@@ -0,0 +1,221 @@
+use std::{
+    collections::BTreeMap,
+    convert::TryInto as _,
+    env, fmt, fs,
+    path::{Path, PathBuf},
+};
+
+use chrono::{Datelike as _, TimeZone as _, Utc};
+use glob::glob;
+use regex::Regex;
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+struct Date {
+    year: u32,
+    month: u32,
+}
+
+impl Date {
+    fn months_since(self, other: Date) -> Option<u32> {
+        let self_chrono = Utc.ymd(self.year.try_into().unwrap(), self.month, 1);
+        let other_chrono = Utc.ymd(other.year.try_into().unwrap(), other.month, 1);
+        let duration_since = self_chrono.signed_duration_since(other_chrono);
+        let days_since = duration_since.num_days();
+        if days_since < 0 {
+            None
+        } else {
+            Some(days_since.try_into().unwrap())
+        }
+    }
+}
+
+impl fmt::Display for Date {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:04}-{:02}", self.year, self.month)
+    }
+}
+
+fn make_date_regex() -> Regex {
+    Regex::new(
+        r"(?x) # insignificant whitespace mode
+        <!--\s*
+        date:\s*
+        (?P<y>\d{4}) # year
+        -
+        (?P<m>\d{2}) # month
+        \s*-->",
+    )
+    .unwrap()
+}
+
+fn collect_dates_from_file(date_regex: &Regex, text: &str) -> Vec<(usize, Date)> {
+    let mut dates = vec![];
+    let mut line = 1;
+    let mut end_of_last_cap = 0;
+    for (byte_range, date) in date_regex.captures_iter(&text).map(|cap| {
+        (
+            cap.get(0).unwrap().range(),
+            Date {
+                year: cap["y"].parse().unwrap(),
+                month: cap["m"].parse().unwrap(),
+            },
+        )
+    }) {
+        for chr in text[end_of_last_cap..byte_range.end].chars() {
+            if chr == '\n' {
+                line += 1;
+            }
+        }
+        dates.push((line, date));
+        end_of_last_cap = byte_range.end;
+    }
+    dates
+}
+
+fn collect_dates(paths: impl Iterator<Item = PathBuf>) -> BTreeMap<PathBuf, Vec<(usize, Date)>> {
+    let date_regex = make_date_regex();
+    let mut data = BTreeMap::new();
+    for path in paths {
+        let text = fs::read_to_string(&path).unwrap();
+        let dates = collect_dates_from_file(&date_regex, &text);
+        if !dates.is_empty() {
+            data.insert(path, dates);
+        }
+    }
+    data
+}
+
+fn filter_dates(
+    current_month: Date,
+    dates_by_file: impl Iterator<Item = (PathBuf, Vec<(usize, Date)>)>,
+) -> impl Iterator<Item = (PathBuf, Vec<(usize, Date)>)> {
+    dates_by_file
+        .map(move |(path, dates)| {
+            (
+                path,
+                dates
+                    .into_iter()
+                    .filter(|(_, date)| {
+                        current_month
+                            .months_since(*date)
+                            .expect("found date that is after current month")
+                            >= 6
```

Comment:
It might be better to pass this in as an argument, or at least make it a `const`, so that it can be changed more easily.

Reply:
Done